modules/javafx.graphics/src/main/java/com/sun/javafx/css/CssUtil.java

```diff
@@ -67,6 +68,25 @@ private CssUtil() {
         CssMetaData[] combined = new CssMetaData[inheritedFromParent.size() + items.length];
         inheritedFromParent.toArray(combined);
         System.arraycopy(items, 0, combined, inheritedFromParent.size(), items.length);
-        return new UnmodifiableArrayList<>(combined, combined.length);
+        return new ImmutableArrayList<>(combined);
+    }
+
+    /** immutable list with random access backed by an array */
+    private static class ImmutableArrayList<T> extends AbstractList<T> implements RandomAccess {
```

Comment:
I'd like to point that a skeletal implementation of `AbstractList` will perform worse than `ArrayList` for any method that needs to walk the list, as the `AbstractList` will use iterators for methods like `indexOf`, `hashCode` and `equals`.  Now that will probably be irrelevant, but as this seems to be a micro optimization, you should be aware of all the trade offs.

Reply:
Good point, thanks!
This also applies to UnmodifiableArrayList.

For completeness sake, I wanted to mention a few issues (not in scope for this PR) that came out of the code review:

- could use `UnmodifiableArrayList` but it stores extra int size.  perhaps a factory method can be added to it for when `size != elements.length`.
- could improve UnmodifiableArrayList with fast(er) `indexOf`, `hashCode`, `equals` per @hjohn 's comment earlier
- `Control.getCssMetaData()` can be improved to skip merging of two lists if skinBase is null
- the same immutable list should be used in `Control.getCssMetaData()` instead of `ArrayList()`

Reply:
I wouldn't do any of the above unless there is a very good reason (and I'm not seeing one).  Just use standard `List.of` as the last step (or use the `Collections.unmodifableList` wrapper); you'll get the most optimized, automatically maintained, bugfree, immutable `List` implementation Java has to offer.  It means another copy will be made; that's fine, this only happens once -- it's not in a hot path.

If you feel like optimizing something, don't bother with `Control.getCssMetaData` either; instead, deduplicate the property lists so there is only one list per unique `Skin` + `Control` combo.  That saves a **complete** list per control **instance**.

Reply:
I will approve this as is, but I agree with John that it would be better to use one of the existing implementations of List: `new UnmodifiableArrayList`, `List.of`, or `Collections.unmodifiableList`. The cost of the extra copy for `List.of` is in the noise compared to the benefit of not having to maintain yet another special case List class. Similarly, the extra word of data storage in `UnmodifiableArrayList` is in the noise. This is a one-per-class not one-per instance list.